### PR TITLE
Submission Team 200 - 2312080_2312105_2312265_2311289_2313429

### PR DESCRIPTION
**Pull request by ngminhhhh**
No  | Tên sinh viên             | Mã số sinh viên | Công việc                        | Tiến độ
--- | ------------------------ | -------------- | -------------------------------- | ------
1   | Nguyễn Công Minh         | 2312080        | Thuật toán Reinforcement Learning | 100%
2   | Phan Huy Quang Minh      | 2312105        | Thuật toán Reinforcement Learning | 100%
3   | Nguyễn Đắc Nghĩa         | 2312265        | Thuật toán Best Fit               | 100%
4   | Ngô Thanh Hùng           | 2311289        | Latex, Đánh giá thuật toán        | 100%
5   | Lê Thành Tiến            | 2313429        | Thuật toán Best Fit               | 100%

# Cutting stock 2D
This repository implements a policy-based system for optimizing stock allocation. The system leverages two distinct algorithms to allocate products to available stocks efficiently:

1. *Best Fit Policy*: A heuristic-based approach that prioritizes placing larger products first to minimize wasted space.
2. *Reinforcement Learning (RL) Policy*: A deep reinforcement learning approach that learns optimal allocation strategies through interaction with the environment.

## Table of Contents
- [Algorithms Overview](#algorithms-overview)
  - [Best Fit Policy](#best-fit-policy)
    - [Algorithm Description](#algorithm-description)
    - [Workflow](#workflow)
  - [Reinforcement Learning Policy](#reinforcement-learning-policy)
    - [Algorithm Description](#algorithm-description-1)
    - [State Representation](#state-representation)
    - [Neural Network Architecture](#neural-network-architecture)
    - [Action Selection](#action-selection)
    - [Reward Function](#reward-function)
    - [Training Process](#training-process)
    - [Advantages and Limitations](#advantages-and-limitations-1)
- [How to Use](#how-to-use)
- [Future Enhancements](#future-enhancements)
- [License](#license)

---

## Algorithms Overview

The system employs two primary algorithms to determine the optimal placement of products within available stocks: the *Best Fit Policy* and the *Reinforcement Learning (RL) Policy*. Each serves different purposes and offers unique advantages depending on the complexity and requirements of the stock allocation problem.

#### Algorithm Description

The *Best Fit Policy* is a heuristic-based algorithm designed to allocate products to stocks by prioritizing the placement of larger products first. The main goal is to minimize wasted space and efficiently utilize available stock areas.

#### Workflow

1. *Product Sorting*:
   - Products are sorted in descending order based on their area (width * height).
   - This ensures that larger products are placed first, reducing the likelihood of wasted space.

2. *Stock Evaluation*:
   - For each product, the algorithm iterates through all available stocks, sorted by their area in descending order.
   - It checks whether the product can fit within the current stock without rotation.

3. *Position Selection*:
   - Within a stock, the algorithm searches for the first available position where the product can be placed.
   - If the product doesn't fit in its original orientation, the algorithm attempts to rotate it (width and height swapped) and checks for placement again.

4. *Placement Decision*:
   - Once a suitable position is found, the product is placed, and the algorithm moves on to the next product.
   - If no suitable position is found in any stock, the product remains unplaced.

#### Algorithm Description

The *Reinforcement Learning (RL) Policy* employs a deep neural network to learn optimal stock allocation strategies through interaction with the environment. Unlike the heuristic-based Best Fit Policy, the RL Policy can adapt and improve its performance based on feedback received from the environment.

#### State Representation

The state represents the current configuration of stocks and the product to be placed. It includes:

- *Stocks Vector*: A flattened representation of all stocks, indicating occupied and available spaces.
- *Product Size*: The dimensions (width, height) of the current product to be placed.
- *Side Features*:
  - *Waste Amount*: Total unused space across all stocks.
  - *Utilization Rate*: Ratio of used space to total available space.

These components are concatenated into a single state vector to provide the neural network with comprehensive information about the current allocation scenario.

#### Neural Network Architecture

The policy network is a simple feedforward neural network implemented using PyTorch, structured as follows:

- *Input Layer*: Size equal to the state vector (STATE_SIZE).
- *Hidden Layers*:
  - *First Hidden Layer*: Fully connected layer with 32 neurons and ReLU activation.
  - *Second Hidden Layer*: Fully connected layer with 32 neurons and ReLU activation.
- *Output Layers*:
  - *Stock Selection Head*: Outputs logits for selecting the appropriate stock (STOCK_SIZE outputs).
  - *X-Coordinate Head*: Outputs logits for the x-coordinate of placement (MAX_X outputs).
  - *Y-Coordinate Head*: Outputs logits for the y-coordinate of placement (MAX_Y outputs).

#### Action Selection

1. *Forward Pass*:
   - The state vector is passed through the neural network to obtain logits for stock selection, x-coordinate, and y-coordinate.

2. *Sampling Actions*:
   - *Stock Selection*: A categorical distribution is created from the stock logits, and a stock index is sampled.
   - *X-Coordinate*: A categorical distribution is created from the x-coordinate logits, and an x-position is sampled.
   - *Y-Coordinate*: A categorical distribution is created from the y-coordinate logits, and a y-position is sampled.

3. *Action Validation*:
   - The sampled position is validated to ensure the product fits within the selected stock without overlapping existing products.
   - If valid, the action is accepted; otherwise, a penalty is applied, and the algorithm searches for alternative placements.

#### Reward Function

The reward function is designed to encourage efficient space utilization and penalize wasted space. It comprises two main components:

- *Waste Minimization*:
  - Calculated as the difference between total stock area and the area occupied by products.
  - Normalized and weighted by a factor alpha.

- *Utilization Rate*:
  - Ratio of used space to total available space.
  - Weighted by a factor beta.

*Reward Calculation*:
```python
reward = -alpha * normalized_waste + beta * normalized_utilization
reward = min(reward, 10.0)
reward += 25  # Base reward
```
#### Testcase:
*Best fit*:

```Total area used: 18608
Episode 0 - Time: 0.2002277374267578
Trials: 24
Total area available: 556481
{'filled_ratio': 0.03, 'trim_loss': 0.31021424271488995}
------------------------------------------------
Total area used: 5506
Episode 1 - Time: 0.19992613792419434
Trials: 16
Total area available: 568571
{'filled_ratio': 0.01, 'trim_loss': 0.40871993127147765}
------------------------------------------------
Total area used: 13158
Episode 2 - Time: 0.200974702835083
Trials: 24
Total area available: 546351
{'filled_ratio': 0.02, 'trim_loss': 0.27312033684441517}
------------------------------------------------
Total area used: 38028
Episode 3 - Time: 0.3127899169921875
Trials: 88
Total area available: 561954
{'filled_ratio': 0.05, 'trim_loss': 0.12112952879103192}
------------------------------------------------
Total area used: 44686
```
*Reinforcement Learning:
```Total area used: 18608
Episode 0 - Time: 0.20075178146362305
Trials: 24
Total area available: 556481
{'filled_ratio': 0.04, 'trim_loss': 0.48276522741101946}
------------------------------------------------
Total area used: 54546
Episode 1 - Time: 0.5184755325317383
Trials: 50
Total area available: 541830
{'filled_ratio': 0.08, 'trim_loss': 0.21650885241723425}
------------------------------------------------
Total area used: 66989
```

### CODE REVIEW

**Comment by martinakaduc** [comment]
Please revise the PR in Dec 16. Otherwise, your group will get 0 score for code.
